pkg/urbit/noun/events.c

```diff
@@ -704,32 +818,85 @@ _ce_patch_apply(u3_ce_patch* pat_u)
+    }                                                                          \
+  } while ( 0 )
+
+#define map_image(image, base_address, page_protections)                       \
+  do {                                                                         \
+    if ( -1 == (c3_ps)mmap(base_address,                                       \
+                    (image)->pgs_w * pag_siz_i,                                \
+                    page_protections,                                          \
+                    MAP_FIXED | MAP_PRIVATE,                                   \
+                    (image)->fid_i,                                            \
+                    0) )                                                       \
+    {                                                                          \
+      fprintf(stderr,                                                          \
+              "loom: failed to map %s snapshot "                               \
+              "image at base address %p: %s\r\n",                              \
+              (image)->nam_c,                                                  \
+              base_address,                                                    \
+              strerror(errno));                                                \
+      exit(1);                                                                 \
+    }                                                                          \
+                                                                               \
+    c3_y* ptr_y = base_address;                                                \
+    for ( c3_w idx_w = 0; idx_w < (image)->pgs_w; idx_w++ ) {                  \
+      mark_page_clean(ptr_y);                                                  \
+      ptr_y += pag_siz_i;                                                      \
+    }                                                                          \
+  } while ( 0 )                                                                \
+
+
+  if ( sou_u && sou_u->pgs_w > 0 ) {
+    c3_y* bas_y = ((c3_y*)u3_Loom + u3a_bytes) - pag_siz_i;
+    read_image(sou_u, bas_y, pro_o, -1);
   }
 
-  c3_w i_w;
-
-  lseek(img_u->fid_i, 0, SEEK_SET);
-  for ( i_w = 0; i_w < img_u->pgs_w; i_w++ ) {
-    c3_assert(-1 != read(img_u->fid_i, ptr_w, pag_siz_i));
-
-    if ( c3y == pro_o ) {
-      c3_assert(0 == mprotect(ptr_w, pag_siz_i, PROT_READ));
+  if ( nor_u && nor_u->pgs_w > 0 ) {
+    if ( c3n == not_map_o ) {
+      const c3_i pro_i = ( c3y == pro_o ) ? PROT_READ : PROT_READ | PROT_WRITE;
+      map_image(nor_u, (c3_y*)u3_Loom, pro_i);
+    }
+    else {
+      read_image(nor_u, (c3_y*)u3_Loom, pro_o, 1);
     }
-
-    c3_w pag_w = u3a_outa(ptr_w) >> u3a_page;
-    c3_w blk_w = pag_w >> 5;
-    c3_w bit_w = pag_w & 31;
-    u3P.dit_w[blk_w] &= ~(1 << bit_w);
-
-    ptr_w += stp_ws;
   }
+#undef mark_page_clean
+#undef read_image
+#undef map_image
```

Comment:
The macros are mostly overkill here: `map_image()` is only performed once, so it doesn't need to be extracted. `read_image()` is a big, expensive i/o operation, so there's no advantage to it being a macro (and the function should use the same argument names and types for clarity). `mark_page_clean()` is a fine macro (although setting up the macro probably uses as many lines as just repeating the bitmap logic).

Reply:
The intent behind all three macros was increased legibility without polluting the namespace of the file with additional static functions. I agree that `map_image()` is overkill, and I've removed it, but `read_image()` is used twice within the function *and* is not used anywhere else in the file, so I'm reluctant to add another function outside of the scope of `_ce_image_apply()`.

Reply:
Can you expound on this? It seems to me that these blocks, whether as macros or static functions, inhabit roughly the same namespace -- the implementation details of this subsystem. Neither impact the external api, and macros have the added strength/difficulty of admitting arbitrary patterns of control flow. And we don't need the optimization of manual inlining for the i/o here (where we arguably want it for the page bitmap).

At the same time, the change I requested here seems to have made the function more ugly -- maybe reversing the branches for handling `north.bin` help?

Reply:
These macros, because they're defined *and* undefined within the same function,
actually occupy an informal, function-local namespace. If they were defined in
the file's namespace (i.e. the static namespace) or, even worse, the global
namespace, they'd be terrible, because any reader of `_ce_image_apply()` would
have to keep in mind the possibility that other functions within the file (or,
even worse, other files), may rely on these macros, which adds cognitive load
for the reader. Admittedly, none of these three macros do anything that a
function can't do. However, because macros can be "scoped" to a particular
function, I'd argue that the slightly higher complexity of macros is justified,
especially because the reader can easily look at the macros and verify that they
don't do anything exotic. Furthermore, they clearly encapsulate the logic of
`_ce_image_apply()`; just by glancing at the following code, it's apparent what
`_ce_image_apply()` does:

```
_ce_image_apply(u3e_image* nor_u, u3e_image* sou_u, c3_o pro_o)
{
#define mark_page_clean(address) /* snip */
#define read_image(image, base_address, should_protect, step_direction) /* snip */
#define map_image(image, base_address, page_protections) /* snip */
  if ( sou_u && sou_u->pgs_w > 0 ) {
    c3_y* bas_y = ((c3_y*)u3_Loom + u3a_bytes) - pag_siz_i;
    read_image(sou_u, bas_y, pro_o, -1);
  }

  if ( nor_u && nor_u->pgs_w > 0 ) {
    if ( u3C.wag_w & u3o_map_snapshot) {
      const c3_i pro_i = ( c3y == pro_o ) ? PROT_READ : PROT_READ | PROT_WRITE;
      map_image(nor_u, (c3_y*)u3_Loom, pro_i);
    }
    else {
      read_image(nor_u, (c3_y*)u3_Loom, pro_o, 1);
    }
  }
#undef mark_page_clean
#undef read_image
#undef map_image
}
```

So the intent here is not performance, but legibility, which I believe is a
legitimate use case of macros. Programming is really just the definition and
subsequent implementation of interfaces stacked one on top of each other 
and nested within each other, and by adding a function-local namespace, we
are able to better express the relationship between parts of an implementation
by only making visible (in this case) to other functions the functionality that
actually concerns them.

Reply:
The code snippet above (with `map_image()`) perfectly encapsulates how I broke the problem down in my head as I was writing the code, which, in addition to correctness, is a property that I'd argue all code should have (or at least as much code as possible).